lib/service/awsoidc.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	issuer, err := awsoidc.IssuerFromPublicAddress(process.proxyPublicAddr().Addr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	clusterNameConfig, err := authClient.GetClusterName()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	clusterName := clusterNameConfig.GetClusterName()
+
+	databases, err := authClient.GetDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	awsRegions := make(map[string]interface{})
+	for _, database := range databases {
+		if database.IsAWSHosted() && database.IsRDS() {
+			awsRegions[database.GetAWS().Region] = nil
+		}
+	}
+
+	var resources []types.Integration
+	var nextKey string
+	for {
+		igs, nextKey, err := authClient.ListIntegrations(ctx, 0, nextKey)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		resources = append(resources, igs...)
+		if nextKey == "" {
+			break
+		}
+	}
+
+	limit := rate.NewLimiter(rate.Every(updateDeployAgentsRateLimit), 1)
+	for _, ig := range resources {
+		spec := ig.GetAWSOIDCIntegrationSpec()
+		if spec == nil {
+			continue
+		}
+		integrationName := ig.GetName()
+
+		for region := range awsRegions {
+			if err := limit.Wait(ctx); err != nil {
+				return trace.Wrap(err)
+			}
+
+			token, err := authClient.GenerateAWSOIDCToken(ctx, types.GenerateAWSOIDCTokenRequest{
+				Issuer: issuer,
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			req := &awsoidc.AWSClientRequest{
+				IntegrationName: ig.GetName(),
+				Token:           token,
+				RoleARN:         spec.RoleARN,
+				Region:          region,
+			}
+
+			deployServiceClient, err := awsoidc.NewDeployServiceClient(ctx, req, authClient)
+			if err != nil {
+				process.log.Warningf("Failed to update deploy service agents: %v", err)
+				continue
+			}
+
+			ownershipTags := map[string]string{
+				types.ClusterLabel:     clusterName,
+				types.OriginLabel:      types.OriginIntegrationAWSOIDC,
+				types.IntegrationLabel: integrationName,
+			}
+
+			// Acquire a lease for the region + integration before attempting to update the deploy service agent.
+			// If the lease cannot be acquired, the update is already being handled by another instance.
+			semLock, err := authClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.SemaphoreKindConnection,
+				SemaphoreName: fmt.Sprintf("update_deploy_service_agents_%s_%s", region, integrationName),
+				MaxLeases:     1,
+				Expires:       process.Clock.Now().Add(updateDeployAgentsInterval),
+				Holder:        "update_deploy_service_agents",
+			})
+
+			if err != nil {
+				if strings.Contains(err.Error(), teleport.MaxLeases) {
+					process.log.Debug("Deploy service agent update is already being processed")
+					continue
+				}
+				return trace.Wrap(err)
+			}
+
+			if err := awsoidc.UpdateDeployServiceAgents(ctx, deployServiceClient, clusterNameConfig.GetClusterName(), teleportVersion, ownershipTags); err != nil {
```

Comment:
Will there be anything in the logs indicating the service is attempting an auto-upgrade?